design/dhcp-less.md

```diff
@@ -0,0 +1,391 @@
+kind: IPPool
+metadata:
+  name: pool-1
+spec:
+  clusterName: cluster
+
+---
+
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPPool
+metadata:
+  name: preprovisioning-pool
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3PreprovisioningDataTemplate
+metadata:
+  name: preprov-data-template
+spec:
+  preprovisioningNetworkData:
+    networks:
+      ipv4:
+      - id: eth0
+        ipAddressFromIPPool: preprovisioning-pool
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3PreprovisioningTemplate
+metadata:
+  name: preprov-template
+spec:
+  template:
+    spec:
+      dataTemplate:
+        name: preprov-data-template
+      hostSelector:
+        matchLabels:
+          pre-provisioning: foo
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3DataTemplate
+metadata:
+  name: data-template
+spec:
+  clusterName: cluster
+  networkData:
+    networks:
+      ipv4:
+      - id: eth0
+        ipAddressFromIPPool: pool-1
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3MachineTemplate
+metadata:
+  name: machine-template
+spec:
+  template:
+    spec:
+      dataTemplate:
+        name: data-template
+      hostSelector:
+        matchLabels:
+          cluster-role: control-plane
+
+```
+
+In this flow we have `preprovisioning-pool` which is not associated with any cluster, this is used to provide an IPAddress during
+the pre-provisioning phase as described above.  To reduce the required size of the pool, the IPClaim will be deleted after the
+preprovisioning phase is completed, e.g the BMH resource becomes available.
+
+In the provisioning phase another pool, associated with a cluster is used to template networkData as in the existing process.
+
+#### Assumptions and Open Questions
+
+TODO
+
+### Inspection on initial registration
+
+On initial registration of a host, inspection is triggered immediately but this process cannot complete without preprovisioning network configuration in a DHCP-less environment (because the IPA ramdisk can't connect back to the Ironic API).
+
+This can be resolved if the BareMetalHost resources are created with the existing [paused annotation](https://github.com/metal3-io/baremetal-operator/blob/main/docs/api.md#pausing-reconciliation), set to a pre-determined value (e.g `metal3.io/preprovisioning`) which can then be removed by the new controller after `preprovisioningNetworkDataName` has been set, then inspection will be able to succeed.
+
+### Implementation Details/Notes/Constraints
+
+#### IP Reuse
+
+A related issue has been previously addressed via the [IP Reuse](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/docs/ip_reuse.md) functionality - this means we can couple IPClaims to the BareMetalHost resources which will enable consistent IP allocations for pre-provisioning and subsequent provisioning operations (provided the same IPPool is used for both steps)
+
```

Comment:
Currently CAPM3 is coupled to the openstack format because it's what Ironic natively supports - perhaps I should clarify in the non-goals but I'm not attempting to change that via this proposal, any optional  support for nmstate as an alternative will have to be considered via another proposal IMO.

Reply:
I'll resolve this for now - let me know if you think we need further clarifications, but my intention with this proposal is to reuse the existing `networkData` templating, adding native support for nmstate is likely a much larger task so can be potentially considered via a future proposal.